synapse/storage/databases/main/user_directory.py

```diff
@@ -811,94 +811,132 @@ async def search_user_dir(
         if isinstance(self.database_engine, PostgresEngine):
             full_query, exact_query, prefix_query = _parse_query_postgres(search_term)
 
-            # If enabled, this config option will rank local users higher than those on
-            # remote instances.
-            if self._prefer_local_users_in_search:
-                # This statement checks whether a given user's user ID contains a server name
-                # that matches the local server
-                statement = "* (CASE WHEN user_id LIKE ? THEN 2.0 ELSE 1.0 END)"
-                additional_ordering_statements.append(statement)
-
-                ordering_arguments += ("%:" + self._server_name,)
-
-            # We order by rank and then if they have profile info
-            # The ranking algorithm is hand tweaked for "best" results. Broadly
-            # the idea is we give a higher weight to exact matches.
-            # The array of numbers are the weights for the various part of the
-            # search: (domain, _, display name, localpart)
-            sql = """
-                SELECT d.user_id AS user_id, display_name, avatar_url
-                FROM user_directory_search as t
-                INNER JOIN user_directory AS d USING (user_id)
-                WHERE
-                    %(where_clause)s
-                    AND vector @@ to_tsquery('simple', ?)
-                ORDER BY
-                    (CASE WHEN d.user_id IS NOT NULL THEN 4.0 ELSE 1.0 END)
-                    * (CASE WHEN display_name IS NOT NULL THEN 1.2 ELSE 1.0 END)
-                    * (CASE WHEN avatar_url IS NOT NULL THEN 1.2 ELSE 1.0 END)
-                    * (
-                        3 * ts_rank_cd(
-                            '{0.1, 0.1, 0.9, 1.0}',
-                            vector,
-                            to_tsquery('simple', ?),
-                            8
-                        )
-                        + ts_rank_cd(
-                            '{0.1, 0.1, 0.9, 1.0}',
-                            vector,
-                            to_tsquery('simple', ?),
-                            8
+            if not search_term:
+                logger.debug(
+                    "search_user_dir() No search term provided to user_directory, doing plain list"
+                )
+                sql = """
+                    SELECT d.user_id AS user_id, display_name, avatar_url
+                    FROM user_directory_search as t
```

Comment:
It doesn't look to me like this table is used anymore in this query; it might be sensible to select directly from `user_directory`.

I wonder if this should be ordered? Of course, there's no ranking information since there's no query, but alphabetical ordering by display name and then user ID seems like it would be sensible (depending on what indices are available)?

There's also a setting, `self._prefer_local_users_in_search`, which this optimised query seems to ignore — optionally influencing the ordering according to a case statement as below might be necessary.